Add support for xdg config folder for docker cli

vdemeester opened the pull request:
Make the docker CLI able to look for `XDG_CONFIG_HOME` for client configuration files. It will do this by default, and thus, on a clean, new docker installation it will always write on `XDG_CONFIG_HOME` (and default is `$HOME/.config`). 

But if a legacy configuration is there (`$HOME/.docker/config.json`), then it will still use it until it is moved to `XDG_CONFIG_HOME`.

This is only for **Linux**. Nothing should change for other platform.

```bash
$ docker info                                              
WARNING: configuration migrated from /root/.docker/config.json to /root/.config/docker/config.json
# […]
$ docker info
# […]
```

- [x] Should we always write to `$XDG_CONFIG_HOME/docker/config.json` or be conservative if the legacy `$HOME/.docker/config.json` is there.
- [x] This will also look at the files in `XDG_CONFIG_DIRS` (and default is `/etc/xdg/` if empty). These might not be writeable so we should probably always write on `XDG_CONFIG_HOME` or the legacy one if existing.
- [ ] Update docs

Reference https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables

Closes #20693

/cc @thaJeztah @icecrime @dnephin @tianon @runcom @cpuguy83 @tiborvass 

🐸 

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

runcom commented:
> But if a legacy configuration is there ($HOME/.docker/config.json), then it will still use it until it is moved to XDG_CONFIG_HOME.

can't we just migrate the old one to `XDG_CONFIG_HOME`?

vdemeester commented:
@runcom that's one of the question ("Should we always write to $XDG_CONFIG_HOME/docker/config.json or be conservative if the legacy $HOME/.docker/config.json is there.") 😝

runcom commented:
> @runcom that's one of the question ("Should we always write to $XDG_CONFIG_HOME/docker/config.json or be conservative if the legacy $HOME/.docker/config.json is there.")

right, my answer was, let's migrate the old config to the new one :smile:

tianon commented:
I'm +1 on the change and on runtime migration 👍

runcom commented:
ping @docker/core-maintainers

thaJeztah commented:
+1 on migrating from the old to the new path

justincormack commented:
Hmm, sorry only just looked at this. I don't think this really exists on OSX, Windows, FreeBSD, Solaris, Linux servers. Maybe it should be conditional based on platform/whether the env vars are set?

vdemeester commented:
> Hmm, sorry only just looked at this. I don't think this really exists on OSX, Windows, FreeBSD, Solaris, Linux servers. Maybe it should be conditional based on platform/whether the env vars are set?

Right, didn't do that in the current state of the PR (but will), but yeah, that would only be *used* for platform that have these, i.e. Linux (server or not).

LK4D4 commented:
@vdemeester looks like people is ok with design. Let's proceed with code.

vdemeester commented:
Updated. Let's see if it's green or not. It's not ready yet, I have to make sure it writes on the right file, always.

LK4D4 commented:
@vdemeester it's not green :D

vdemeester commented:
damn 😓

justincormack commented:
@vdemeester so long as you do nothing on non Linux and nothing on Linux if `XDG_CONFIG_HOME` is unset thats fine.

vdemeester commented:
> nothing on Linux if `XDG_CONFIG_HOME` is unset thats fine.

why for this case ?

vdemeester commented:
Hum ok, so I have to handle that for tls and trust files too (the "migration").. The code to change is gonna be a bit bigger than I thought..

vdemeester committed:
Add support for xdg config folder for docker

Make the docker CLI able to look for `XDG_CONFIG_HOME` for client
configuration files. It will do this by default, and thus, on a clean,
new docker installation it will always write on `XDG_CONFIG_HOME` (and
default is `$HOME/.config`).

But if a legacy configuration is there (`$HOME/.docker/config.json`),
then it will still use it until it is moved to `XDG_CONFIG_HOME`.

This will only work on Linux. Any other platform should be left
unchanged.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

tiborvass commented:
@vdemeester I agree with https://github.com/docker/docker/pull/30025#issuecomment-276107506

Please rebase and let's get this merged :)

tianon commented:
If we do nothing when `XDG_CONFIG_HOME` is unset, then this change probably won't do much for the majority of users and isn't worth making anymore IMO (my own systems don't have that set, and I'm using a fully graphical environment in a pretty stock configuration on both Debian and Gentoo).

My own previous objections to using `~/.config` were that we didn't have a lot of examples of non-X applications using it, and there are at least a couple now that I can think of off the top of my head:

- `~/.config/htop`
- `~/.config/syncthing`

vdemeester commented:
@justincormack @tiborvass I do agree with @tianon *"If we do nothing when XDG_CONFIG_HOME is unset, then this change probably won't do much for the majority of users and isn't worth making anymore IMO"*.

thaJeztah commented:
Orthogonal to this change, but I recall discussing with @vdemeester; is there a convention where to use _sensitive_ data (e.g. authentication data)? The `config.json` allows users to customize docker (for example, custom formatting for `docker ps`). I can see it useful if users are able to put their config in a "dot files" repository, but given that the `config.json` can also contain credentials (if no credentials-helper is used), that's not possible.

tianon commented:
Maybe we could move those to a separate "auth.json" with appropriate file
modes, similar to files like "id_rsa" ?  I'm not aware of anything else in
XDG (or prevailing convention) for handling security-sensitive
configuration differently.

vdemeester commented:
> Maybe we could move those to a separate "auth.json" with appropriate file modes, similar to files like "id_rsa" ? I'm not aware of anything else in XDG (or prevailing convention) for handling security-sensitive configuration differently.

I defenitely want to extract the auth part elsewhere yep 👼

derimagia commented:
"do nothing on non Linux"

Can we support this on mac as well? Don't see why it should be ignored there as well. The only time it's set is when people want it to move.

tiborvass commented:
Ping @vdemeester

vdemeester commented:
ping @dnephin @thaJeztah @tiborvass @tianon what should we do ? If we go for :don't do anything if `XDG_CONFIG_DIR` is not set, I'll just close because it make the code more complex and won't be used that much (… and will be more confusing to user and to document than anything)

vdemeester commented:
We discussed it on the Maintainer session. We need to discuss more on the related issue (#20693), mainly about what to do when migrating, what to do if there is multiple configuration files (old path, new path) and if this should work only on linux, and/or only if `XDG_CONFIG_DIR` is set.

I'll close this for now. Let's continue discussing it on #20693.
